addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,65 +1,124 @@
+        end_date = date_utils.end_of(date_utils.end_of(from_date, 'month'), 'day')
+
         select_query = """
-SELECT user_id,
-(
-    SELECT old_value from gamification_karma_tracking old_tracking
-    WHERE old_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date ASC LIMIT 1
-), (
-    SELECT new_value from gamification_karma_tracking new_tracking
-    WHERE new_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date DESC LIMIT 1
-)
-FROM gamification_karma_tracking
-WHERE tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-AND consolidated IS NOT TRUE
-GROUP BY user_id """
+        WITH old_tracking AS (
+            SELECT DISTINCT ON (user_id) user_id, old_value, tracking_date
+              FROM gamification_karma_tracking
+             WHERE tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND consolidated IS NOT TRUE
+          ORDER BY user_id, tracking_date ASC, id ASC
+        )
+            SELECT DISTINCT ON (nt.user_id)
+                            nt.user_id,
+                            ot.old_value AS old_value,
+                            nt.new_value AS new_value,
+                            ot.tracking_date AS from_tracking_date,
+                            nt.tracking_date AS to_tracking_date
+              FROM gamification_karma_tracking AS nt
+              JOIN old_tracking AS ot
+                   ON ot.user_id = nt.user_id
+             WHERE nt.tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND nt.consolidated IS NOT TRUE
+          ORDER BY nt.user_id, nt.tracking_date DESC, id DESC
+        """
+
         self.env.cr.execute(select_query, {
             'from_date': from_date,
-            'to_date': end_date,
+            'end_date': end_date,
         })
         results = self.env.cr.dictfetchall()
         if results:
-            for result in results:
-                result['consolidated'] = True
-                result['tracking_date'] = fields.Date.to_string(from_date)
-            self.create(results)
+            self.create([{
+                'consolidated': True,
+                'new_value': result['new_value'],
+                'old_value': result['old_value'],
+                'origin_ref': f'res.users,{self.env.user.id}',
+                'reason': _(
+                    'Consolidation from %s to %s',
+                    result['from_tracking_date'].strftime('%Y-%m-%d'),
+                    result['to_tracking_date'].strftime('%Y-%m-%d'),
+                ),
+                'tracking_date': fields.Datetime.to_string(from_date),
+                'user_id': result['user_id'],
+            } for result in results])
```

Comment:
So now I'm wondering why we add all those reason and other detail, if at the end, we will lose all the granularity after the consolidation. We want details BUT only for 2 months :shrug: